src/geode/basic/database.cpp

```diff
@@ -0,0 +1,345 @@
+        }
+
+        const uuid& register_unique_data( std::unique_ptr< Identifier >&& data )
+        {
+            const auto& registedred_data = register_data( std::move( data ) );
+            save_data( registedred_data );
+            return registedred_data->id();
+        }
+
+        std::shared_ptr< Storage > data( const uuid& id ) const
+        {
+            auto& storage = storage_.at( id );
+            if( !storage || storage->expired() )
+            {
+                storage = std::make_shared< Storage >( load_data( id ) );
+            }
+            return storage;
+        }
+
+        std::unique_ptr< Identifier > take_data( const uuid& id )
+        {
+            auto& storage = storage_.at( id );
+            if( storage && storage->unused() && !storage->expired() )
+            {
+                auto* data = storage->take_data();
+                storage.reset();
+                return std::unique_ptr< Identifier >{ data };
+            }
+            return load_data( id );
+        }
+
+        void delete_data( const uuid& id )
+        {
+            storage_.erase( id );
+        }
+
+        void register_serializer_functions(
+            serializer_function serializer, serializer_function deserializer )
+        {
+            serializers_.push_back( serializer );
+            deserializers_.push_back( deserializer );
+        }
+
+    private:
+        const std::unique_ptr< Identifier >& register_data(
+            std::unique_ptr< Identifier >&& data )
+        {
+            const auto& id = data->id();
+            const auto it = storage_.find( id );
+            if( it != storage_.end() )
+            {
+                if( it->second->unused() )
+                {
+                    it->second->set_data( std::move( data ) );
+                    return it->second->data();
+                }
+                delete_data( id );
+                return do_register_data( std::move( data ) );
+            }
+            return do_register_data( std::move( data ) );
+        }
+
+        const std::unique_ptr< Identifier >& do_register_data(
+            std::unique_ptr< Identifier >&& data )
+        {
+            const auto id = data->id();
+            auto new_storage = std::make_shared< Storage >( std::move( data ) );
+            return storage_.emplace( id, std::move( new_storage ) )
+                .first->second->data();
+        }
+
+        void save_data( const std::unique_ptr< Identifier >& data ) const
+        {
+            const auto filename =
+                absl::StrCat( directory_, "/", data->id().string() );
+            std::ofstream file{ filename, std::ofstream::binary };
+            TContext context;
+            for( const auto& serializer : serializers_ )
+            {
+                serializer( std::get< 0 >( context ) );
+            }
+            Serializer archive{ context, file };
+            archive.ext( data, bitsery::ext::StdSmartPtr{} );
```

Comment:
This can only be done for Bitsery-compatible objects. I think we could add this as mandatory. We do this database for internal usage. If yes, some of our objects need better bitsery support like BRep but it is not a problem

Reply:
you mean BRep or BMesh?

Reply:
BRep, but both are not fully compatible at the moment. There is no serialize method.